### PR TITLE
Np 1302 limit access elasticsearch

### PR DESCRIPTION
**Pull request by svenngjeto**
DynamoDB trigger is moved in from nva-elasticsearch-dataloader and uses the same document classes. Standard jva 11 HttpClient is replaced with Apache http client with signing to make fine-grained access control possible.

### CODE REVIEW

File: search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        }
+    }
+
+    private boolean isNotValidEventName(String eventName) {
+        return !(eventName.equals(INSERT) || eventName.equals(MODIFY) || eventName.equals(REMOVE));
+    }
+
+    private void upsertSearchIndex(DynamodbEvent.DynamodbStreamRecord streamRecord)
+            throws SearchException {
+        Map<String, AttributeValue> valueMap = streamRecord.getDynamodb().getNewImage();
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java:114]: Extract this as "logStreamRecord"

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+     * @param document the document to be inserted
+     * @throws SearchException when something goes wrong
+     * */
+    public void addDocumentToIndex(IndexDocument document) throws SearchException {
+
+        logger.debug(UPSERTING_LOG_MESSAGE, document);
+
+        try {
+            String jsonDocument = document.toJsonString();
+            IndexRequest indexRequest = new IndexRequest(elasticSearchEndpointIndex)
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:133]: Possible to extract 133-137 to doUpsert method?

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+     * Removes an document from Elasticsearch index.
+     * @param identifier og document
+     * @throws SearchException when
+     */
+    public void removeDocumentFromIndex(String identifier) throws SearchException {
+        logger.trace(DELETE_LOG_MESSAGE, identifier);
+
+        try {
+
+            DeleteRequest deleteRequest = new DeleteRequest(elasticSearchEndpointIndex, identifier);
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:155]: doDelete method?

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    private Stream<JsonNode> toStream(JsonNode node) {
+        return StreamSupport.stream(node.spliterator(), false);
+    }
+
+    protected final RestHighLevelClient createElasticsearchClientWithInterceptor(String serviceName,
+                                                                                 String region,
+                                                                                 String elasticSearchEndpoint) {
+        AWS4Signer signer = new AWS4Signer();
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:191]: Extract signing to own method

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java (hunk 1 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java]: Extracted

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java (hunk 2 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java]: Extracted

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java (hunk 3 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java]: Extracted

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java (hunk 4 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java]: Extracted

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 5]
```diff
@@ -105,20 +106,25 @@ public ElasticSearchHighLevelRestClient(Environment environment, RestHighLevelCl
...
+        final SearchSourceBuilder sourceBuilder = new SearchSourceBuilder()
+            .query(queryBuilder)
+            .size(results);
+        final SearchRequest searchRequest = new SearchRequest(elasticSearchEndpointIndex);
+        searchRequest.source(sourceBuilder);
+        logger.debug(SEARCH_REQUEST_MESSAGE, elasticSearchEndpointIndex, term, searchRequest);
+        SearchResponse searchResponse = elasticSearchClient.search(searchRequest, RequestOptions.DEFAULT);
+        return searchResponse;
+    }
+
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:127]: Possibly (although naming is weak)

```

   private SearchResponse doSearch(String term, int results) throws IOException {
        final SearchRequest searchRequest = new SearchRequest(elasticSearchEndpointIndex);
        searchRequest.source(getSearchSourceBuilder(term, results));
        logSearch(term, searchRequest);
        return elasticSearchClient.search(searchRequest, RequestOptions.DEFAULT);
    }

    private void logSearch(String term, SearchRequest searchRequest) {
        logger.debug(SEARCH_REQUEST_MESSAGE, elasticSearchEndpointIndex, term, searchRequest);
    }

    private SearchSourceBuilder getSearchSourceBuilder(String term, int results) {
        return new SearchSourceBuilder()
                .query(queryStringQuery(term))
                .size(results);
    }
```

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java (hunk 5 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:127]: extraced searchRequest ganeration, removed logging without interest anymore

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 6]
```diff
@@ -114,15 +110,19 @@ public SearchResourcesResponse searchSingleTerm(String term, int results) throws
     }
 
     private SearchResponse doSearch(String term, int results) throws IOException {
+        final SearchRequest searchRequest = getSearchRequest(term, results);
+        SearchResponse searchResponse = elasticSearchClient.search(searchRequest, RequestOptions.DEFAULT);
+        return searchResponse;
+    }
+
+    private SearchRequest getSearchRequest(String term, int results) {
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:118]: maybe just inline the static import of queryStringQuery

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 7]
```diff
@@ -142,13 +139,17 @@ public void addDocumentToIndex(IndexDocument document) throws SearchException {
...
+
+    private UpdateRequest getUpdateRequest(IndexDocument document) throws JsonProcessingException {
         IndexRequest indexRequest = new IndexRequest(elasticSearchEndpointIndex)
                 .source(document.toJsonString(), XContentType.JSON);
         UpdateRequest updateRequest = new UpdateRequest(elasticSearchEndpointIndex,  document.getIdentifier())
             .upsert(indexRequest)
             .doc(indexRequest);
-        logger.debug(ADD_DOCUMENT_LOG_MESSAGE, elasticSearchEndpointIndex, document.getIdentifier());
-        elasticSearchClient.update(updateRequest, RequestOptions.DEFAULT);
+        return updateRequest;
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:152]: inline

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java [hunk 8]
```diff
@@ -114,15 +110,19 @@ public SearchResourcesResponse searchSingleTerm(String term, int results) throws
...
         QueryStringQueryBuilder queryBuilder = QueryBuilders.queryStringQuery(term);
         final SearchSourceBuilder sourceBuilder = new SearchSourceBuilder()
             .query(queryBuilder)
             .size(results);
         final SearchRequest searchRequest = new SearchRequest(elasticSearchEndpointIndex);
         searchRequest.source(sourceBuilder);
-        logger.debug(SEARCH_REQUEST_MESSAGE, elasticSearchEndpointIndex, term, searchRequest);
-        SearchResponse searchResponse = elasticSearchClient.search(searchRequest, RequestOptions.DEFAULT);
-        return searchResponse;
+        return searchRequest;
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:125]: inline

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java (hunk 8 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java]: inlined

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java (hunk 6 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java:116]: inlined

File: search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java (hunk 7 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/search/ElasticSearchHighLevelRestClient.java]: inlined

File: search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java [hunk 9]
```diff
@@ -85,10 +85,16 @@ private void processRecord(DynamodbEvent.DynamodbStreamRecord streamRecord) thro
         validate(streamRecord);
         String eventName = streamRecord.getEventName();
 
-        if (eventName.equals(INSERT) || eventName.equals(MODIFY)) {
+        if (!eventName.equals(INSERT)) {
```
Comment (brinxmat) [inline search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java:88]: Revert to 
```
        if (eventName.equals(INSERT) || eventName.equals(MODIFY)) {
            upsertSearchIndex(streamRecord);
        } else if (eventName.equals(REMOVE)) {
            removeFromSearchIndex(streamRecord);
        }
```

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 10]
```diff
@@ -67,6 +67,7 @@ subprojects {
         testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: junit5Version
         testImplementation group: 'org.junit.jupiter', name: 'junit-jupiter-api', version: junit5Version
         testRuntimeOnly group: 'org.junit.jupiter', name: 'junit-jupiter-engine', version: junit5Version
+        testCompile group: 'org.junit.jupiter', name: 'junit-jupiter-params', version: junit5Version
```
[the remark this answers is not in the document]
Comment (svenngjeto) [reply build.gradle:70]: It's not.

File: search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java (hunk 9 above)
Comment (svenngjeto) [reply search-resources-api/src/main/java/no/unit/nva/dynamodb/DynamoDBStreamHandler.java:88]: Reverted

**Review by brinxmat — APPROVED** [review]

**Pull request merged**